### PR TITLE
Error handling cleanup in solvers/flattening 3

### PR DESCRIPTION
**Pull request by danpoe**

### CODE REVIEW

File: src/solvers/flattening/boolbv_mult.cpp [hunk 1]
```diff
@@ -12,6 +12,8 @@ Author: Daniel Kroening, kroening@kroening.com
 
 bvt boolbvt::convert_mult(const exprt &expr)
 {
+  PRECONDITION(expr.id() == ID_mult || expr.id() == "no-overflow-mult");
```
Comment (danpoe) [inline src/solvers/flattening/boolbv_mult.cpp]: Is `"no-overflow-mult"` still used? It does not seem to be assigned anywhere. If not, we can remove it and make `convert_mult` take a `mult_exprt` instead.

File: src/solvers/flattening/boolbv_overflow.cpp [hunk 2]
```diff
@@ -173,18 +174,23 @@ literalt boolbvt::convert_overflow(const exprt &expr)
   else if(has_prefix(expr.id_string(), "overflow-typecast-"))
```
Comment (danpoe) [inline src/solvers/flattening/boolbv_overflow.cpp]: Is `"overflow-typecast-"` still used? It's not assigned anywhere.

File: src/solvers/flattening/boolbv_onehot.cpp [hunk 3]
```diff
@@ -11,6 +11,8 @@ Author: Daniel Kroening, kroening@kroening.com
 
 literalt boolbvt::convert_onehot(const unary_exprt &expr)
 {
+  PRECONDITION(expr.id() == ID_onehot || expr.id() == ID_onehot0);
```
Comment (danpoe) [inline src/solvers/flattening/boolbv_onehot.cpp:14]: Are those expressions used anywhere? The IDs are not assigned anywhere.

File: src/solvers/flattening/boolbv_mult.cpp [hunk 4]
```diff
@@ -21,39 +23,36 @@ bvt boolbvt::convert_mult(const exprt &expr)
...
   const exprt::operandst &operands=expr.operands();
-  if(operands.empty())
-    throw "mult without operands";
+  DATA_INVARIANT(
+    !operands.empty(),
+    "multiplication must have operands");
 
   const exprt &op0=expr.op0();
 
   bool no_overflow=expr.id()=="no-overflow-mult";
```
Comment (peterschrammel) [inline src/solvers/flattening/boolbv_mult.cpp]: This can be removed then too.

File: src/solvers/flattening/boolbv_onehot.cpp [hunk 5]
```diff
@@ -25,8 +27,9 @@ literalt boolbvt::convert_onehot(const unary_exprt &expr)
 
   if(expr.id()==ID_onehot)
     return prop.land(one_seen, !more_than_one_seen);
-  else if(expr.id()==ID_onehot0)
-    return !more_than_one_seen;
   else
-    throw "unexpected onehot expression";
+  {
+    INVARIANT(expr.id() == ID_onehot0, "");
```
Comment (peterschrammel) [inline src/solvers/flattening/boolbv_onehot.cpp]: message missing

File: src/solvers/flattening/boolbv_onehot.cpp (hunk 3 above)
Comment (peterschrammel) [reply src/solvers/flattening/boolbv_onehot.cpp:14]: They are used in ebmc.

File: src/solvers/flattening/boolbv_overflow.cpp [hunk 6]
```diff
@@ -173,18 +174,23 @@ literalt boolbvt::convert_overflow(const exprt &expr)
...
   {
     std::size_t bits=unsafe_string2unsigned(id2string(expr.id()).substr(18));
+    INVARIANT(
+      bits != 0,
+      "");
```
Comment (peterschrammel) [inline src/solvers/flattening/boolbv_overflow.cpp]: missing message

File: src/solvers/flattening/boolbv_overflow.cpp [hunk 7]
```diff
@@ -173,18 +174,23 @@ literalt boolbvt::convert_overflow(const exprt &expr)
...
 
     const exprt &op=operands[0];
 
     const bvt &bv=convert_bv(op);
 
-    if(bits>=bv.size() || bits==0)
-      throw "overflow-typecast got wrong number of bits";
+    INVARIANT(
+      bits < bv.size(),
+      "");
```
Comment (peterschrammel) [inline src/solvers/flattening/boolbv_overflow.cpp]: missing message

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 08b7bc06c49534c0bd3580dfb7d22e5264f2ef34).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85552876

File: src/solvers/flattening/boolbv_onehot.cpp (hunk 3 above)
Comment (danpoe) [reply src/solvers/flattening/boolbv_onehot.cpp:14]: Raised an issue for those as well: #3040

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: d57737aff1745ef5d48fcf97f7f4b4e30ea39b51).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85868517

File: src/solvers/flattening/boolbv_overflow.cpp [hunk 8]
```diff
@@ -173,18 +173,19 @@ literalt boolbvt::convert_overflow(const exprt &expr)
...
+      operands.size() == 1,
+      "expression " + expr.id_string() + " should have one operand");
 
     const exprt &op=operands[0];
 
     const bvt &bv=convert_bv(op);
 
-    if(bits>=bv.size() || bits==0)
-      throw "overflow-typecast got wrong number of bits";
+    INVARIANT(bits < bv.size(), "");
```
Comment (tautschnig) [inline src/solvers/flattening/boolbv_overflow.cpp]: Missing message

File: src/solvers/flattening/boolbv_overflow.cpp [hunk 9]
```diff
@@ -173,18 +173,19 @@ literalt boolbvt::convert_overflow(const exprt &expr)
   else if(has_prefix(expr.id_string(), "overflow-typecast-"))
   {
     std::size_t bits=unsafe_string2unsigned(id2string(expr.id()).substr(18));
+    INVARIANT(bits != 0, "");
```
Comment (tautschnig) [inline src/solvers/flattening/boolbv_overflow.cpp]: Missing message

**Comment by danpoe** [comment]
@tautschnig @peterschrammel I think this is ready to be merged.

**Review by tautschnig — APPROVED** [review]

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**